File: 130. Surrounded Regions.cpp
// https://leetcode.com/problems/surrounded-regions/

// Given an m x n matrix board containing 'X' and 'O', capture all regions
// surrounded by 'X'.

// A region is captured by flipping all 'O's into 'X's in that surrounded region.

////////////////////////////////////////////////////////////////////////////////

// dfs to grow from edges, filp the rest
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        m = board.size(); n = board[0].size();
        // dfs from edges, mark 'O' to 'S'
        for (int i = 0; i < m; ++i) {
            if (board[i][0] == 'O') dfs(i, 0, board);
            if (board[i][n - 1] == 'O') dfs(i, n - 1, board);
        }
        for (int j = 0; j < n; ++j) {
            if (board[0][j] == 'O') dfs(0, j, board);
            if (board[m - 1][j] == 'O') dfs(m - 1, j, board);
        }
        // mark 'O' to 'X' and 'S' to 'O';
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == 'S') board[i][j] = 'O';
            }
        }
        return;
    }
private:
    int m, n;
    vector<pair<int, int>> dirs = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    void dfs(int i, int j, vector<vector<char>>& board) {
        board[i][j] = 'S';
        for (const auto& [iIncr, jIncr] : dirs) {
            int iNext = i + iIncr, jNext = j + jIncr;
            if (iNext >= 0 && iNext < m && jNext >= 0 && jNext < n
                && board[iNext][jNext] == 'O') {
                dfs(iNext, jNext, board);
            }
        }
        return;
    }
};

File: 130. Surrounded Regions.py
# https://leetcode.com/problems/surrounded-regions/

# Given an m x n matrix board containing 'X' and 'O', capture all regions
# surrounded by 'X'.

# A region is captured by flipping all 'O's into 'X's in that surrounded region.

################################################################################

# dfs start from edges to mark connected 'O' -> 'S'
# change 'O' -> 'X' and 'S' -> 'O

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        if not board or len(board) == 1 or len(board[0]) == 1: return
        
        self.m, self.n = len(board), len(board[0])
        self.board = board
        self.dirs = [(-1,0), (1,0), (0,-1), (0,1)]
        
        # dfs start from edges
        for i in range(self.m):
            if self.board[i][0] == "O": # first col
                self.dfs(i, 0)
            if self.board[i][self.n - 1] == "O": # last col
                self.dfs(i, self.n - 1)
        for j in range(self.n):
            if self.board[0][j] == "O": # first row
                self.dfs(0, j)
            if self.board[self.m - 1][j] == "O": # last row
                self.dfs(self.m - 1, j)
        
        # change 'O' -> 'X' and 'S' -> 'O'
        for i in range(self.m):
            for j in range(self.n):
                if self.board[i][j] == 'O':
                    self.board[i][j] = 'X'
                elif self.board[i][j] == 'S':
                    self.board[i][j] = 'O'
        return
    
    def dfs(self, i, j):
        self.board[i][j] = "S"
        for dir_idx in range(4):
            i_next = i + self.dirs[dir_idx][0]
            j_next = j + self.dirs[dir_idx][1]
            if (0 <= i_next < self.m and 0 <= j_next < self.n
                and self.board[i_next][j_next] == "O"):
                self.dfs(i_next, j_next)
